parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java

```diff
@@ -1254,14 +1276,41 @@ static FileMetaData filterFileMetaDataByMidpoint(FileMetaData metaData, RangeMet
     return metaData;
   }
 
+  private static long tryUseFileOffset(RowGroup rowGroup, long preStartIndex, long preCompressedSize) {
+    long startIndex = 0;
+    startIndex = rowGroup.getFile_offset();
+    // skip checking the first rowGroup
+    // (in case of summary file, there are multiple first groups from different footers)
+    if (preStartIndex != 0 && preStartIndex <= startIndex) {
+
+      //calculate start index for other blocks
+      long minStartIndex = preStartIndex + preCompressedSize;
+      if (startIndex < minStartIndex) {
+        // a bad offset detected, try first column's offset
+        // can not use minStartIndex in case of padding
+        startIndex = getOffset(rowGroup.getColumns().get(0));
```

Comment:
this will throw an exception in encrypted files.

Reply:
In case of  encrypted files with wongly setted file_offset, i have no idea how to fix it when alignment padding take place.
If no padding, then we can just use the calculated index.
I didn't find any footer meta about padding position or something else indicating padding or not.

Reply:
got it, will check a few things

Reply:
@loudongfeng @gszadovszky @shangxinli @sunchao . Looks like the problem is only in one mode of encryption ("encrypted footer", where the ColumnMetaData is not set). Let me propose the following:


      long minStartIndex = preStartIndex + preCompressedSize;
      if (startIndex < minStartIndex) {
        // a bad offset detected, try first column's offset if available
       ColumnChunk columnChunk = rowGroup.getColumns().get(0);
       if (columnChunk.isSetMeta_data()) {
         startIndex = getOffset(columnChunk);
       } else { // EncryptedFooter mode. Plaintext ColumnMetaData is not available.
         // use minStartIndex (imprecise in case of padding, but good enough for filtering)
         startIndex = minStartIndex;
      }


what do you think?

Reply:
that might work for range filtering. as for the explicit offset filtering, the precision might be important. In systems that [use padding](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/HadoopOutputFile.java#L37) (HDFS, WEBHDFS, VIEWFS)  - instead of `startIndex = minStartIndex` , we might have to throw an exception when this is called from  `filterFileMetaDataByStart` saying that this file can't be split by offsets, and should be processed without splitting. Again, given the number of conditions for this situation to occur, and the fact that 1.12.0 is not released yet any framework, such exceptions might never be thrown in reality.

Reply:
Yep, that should be it (IMO). This will work for all files written by 1.12.1 (naturally), for all regular/unencrypted files files written by 1.12.0 and for most (in practice, for all) of encrypted files written by 1.12.0. Tomorrow, I'll send an update to the snippet above.

Reply:
@ggershinsky, your proposal sounds perfect to me.Looking forward to your patch, or shall I update commit following your proposal?

Reply:
Another choice: suppose columnChunk.isSetMeta_data() is the same across different row groups, how about using first columns's offset by default, only using file offset with "encrypted footer"?
(And only throw exception when bad file offset detected and called from filterFileMetaDataByStart, as you suggested.)
@ggershinsky

Reply:
Thanks @loudongfeng, sure, please update the pr with these changes. Regarding columnChunk.isSetMeta_data() - the result will be the same across different row groups, since the whole file is written in one mode (either with the ColumnMetaData, or without them). If they are unavailable, we can't use them for splits, so will have to throw an exception in case of filterFileMetaDataByStart (only for bad columns where startIndex < minStartIndex)

Reply:
and of course, I'm fine with your proposal to change the order - if you can detect the "encrypted footer" mode (eg by checking one columnChunk.isSetMeta_data()), then its ok to start with using first columns's offset

Reply:
Commit updated following ggershinsky's suggestions.Thanks.